keps/sig-network/20191020-default-deny-pod-traffic

```diff
@@ -0,0 +1,270 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Make pod-to-pod communication in Kubernetes secure by default.
+Pods should require an explicit NetworkPolicy ingress configuration,
+in order to receive traffic.
+
+This proposal explicitly targets ingress traffic.
+Egress traffic is not _currently_ being targeted,
+not is not yet deemed out of scope.
+
+## Motivation
+
+Kubernetes and Kubernetes workloads are notoriously easy to leave insecure.
+Many users are not aware of relevant settings (like NetworkPolicy),
+or choose not to configure them due to friction and learning curves.
+This is a systemic pattern,
+NetworkPolicy is one example.
+
+TODO pull examples
+
+### Goals
+
+* Restrict pod ingress by default on a cluster.
+
+### Non-Goals
+
+What is out of scope for this KEP?
+Listing non-goals helps to focus discussion and make progress.
+
+## Proposal
+
+There are multiple options for design/implementation.
+
+Restricting network access requires a CNI driver that supports NetworkPolict.
+All options will require codifying NetworkPolicy support as a requirement,
+or creating a clear user experience when using a CNI driver that does not support NetworkPolicy.
+
+### Option 1: Change default NetworkPolicy behavior
+
+The most conceptually simple option is to change default NetworkPolicy behavior.
+Currently,
+NetworkPolicy allows all traffic to pods that are not selected by a NetworkPolicy. [1]
+
+If all pods are treated as isolated,
+users must explicitly create NetworkPolicy objects
+(either as needed, or by allowing all ingress and _explicitly_ enabling the existing behavior).
+This would potentially require a v2 API, due to changes in behavior (if not the API itself).
```

Comment:
This isn't a change in the behavior of NetworkPolicy though, it's a change in the behavior of lack-of-NetworkPolicies. Changing the API version of NetworkPolicy would not make that non-breaking.

The original design docs seem to have gotten lost in some website/repo reorg a few years back, but we originally specified that in the absence of any NetworkPolicies, the default behavior was undefined. It is definitely allowable for a network plugin to say that pods are isolated by default. (eg, see the pseudo-code [here](https://github.com/kubernetes/kubernetes/pull/24154/files#diff-d07dd9aa199f30808c8d5d44f2f6135fR218); this is in terms of the pre-v1 API where there was an explicit per-Namespace default-deny annotation, but you can see that the expectation was that the default might not always be allow.)

Reply:
Thanks, that's some interesting context.

Given that, the option of creating policies automatically makes more sense than actually changing the default behaviour.